cmd/kubeadm/app/phases/controlplane/manifests.go

```diff
@@ -129,6 +111,74 @@ func CreateStaticPodFiles(manifestDir, patchesDir string, cfg *kubeadmapi.Cluste
+		LivenessProbe:   staticpodutil.LivenessProbe(staticpodutil.GetAPIServerProbeAddress(endpoint), "/livez", int(endpoint.BindPort), v1.URISchemeHTTPS),
+		ReadinessProbe:  staticpodutil.ReadinessProbe(staticpodutil.GetAPIServerProbeAddress(endpoint), "/readyz", int(endpoint.BindPort), v1.URISchemeHTTPS),
+		StartupProbe:    staticpodutil.StartupProbe(staticpodutil.GetAPIServerProbeAddress(endpoint), "/livez", int(endpoint.BindPort), v1.URISchemeHTTPS, cfg.APIServer.TimeoutForControlPlane),
+		Resources:       staticpodutil.ComponentResources("250m"),
+		Env:             kubeadmutil.GetProxyEnvVars(),
+	}, mounts.GetVolumes(kubeadmconstants.KubeAPIServer),
+		map[string]string{kubeadmconstants.KubeAPIServerAdvertiseAddressEndpointAnnotationKey: endpoint.String()})
+
+	if features.Enabled(cfg.FeatureGates, features.RootlessControlPlane) {
+		controlPlanUsersAndGroups, err := GetUserAndGroupIDs()
+		if err != nil {
+			return pod, err
+		}
+		uid := controlPlanUsersAndGroups[kubeadmconstants.KubeAPIServerUserName]
+		if !isDryRun {
+			saPublicKeyFile := filepath.Join(cfg.CertificatesDir, kubeadmconstants.ServiceAccountPublicKeyName)
+			if err := UpdateFileOwnership(saPublicKeyFile, uid, uid, 0600); err != nil {
+				return pod, err
+			}
+
+			saPrivateKeyFile := filepath.Join(cfg.CertificatesDir, kubeadmconstants.ServiceAccountPrivateKeyName)
+			if err := UpdateFileOwnership(saPrivateKeyFile, 0, controlPlanUsersAndGroups[kubeadmconstants.ServiceAccountKeyReadersGroupName], 0640); err != nil {
+				return pod, err
+			}
+
+			apiServerKeyFile := filepath.Join(cfg.CertificatesDir, kubeadmconstants.APIServerKeyName)
+			if err := UpdateFileOwnership(apiServerKeyFile, uid, uid, 0600); err != nil {
+				return pod, err
+			}
+
+			apiServerKubeletClientKeyFile := filepath.Join(cfg.CertificatesDir, kubeadmconstants.APIServerKubeletClientKeyName)
+			if err := UpdateFileOwnership(apiServerKubeletClientKeyFile, uid, uid, 0600); err != nil {
+				return pod, err
+			}
+
+			frontProxyClientKeyName := filepath.Join(cfg.CertificatesDir, kubeadmconstants.FrontProxyClientKeyName)
+			if err := UpdateFileOwnership(frontProxyClientKeyName, uid, uid, 0600); err != nil {
+				return pod, err
+			}
+
+			if cfg.Etcd.External == nil {
+				apiServerEtcdClientKeyFile := filepath.Join(cfg.CertificatesDir, kubeadmconstants.APIServerEtcdClientKeyName)
+				if err := UpdateFileOwnership(apiServerEtcdClientKeyFile, uid, uid, 0600); err != nil {
+					return pod, err
+				}
+			}
+		}
+		pod.Spec.Containers[0].SecurityContext = &v1.SecurityContext{
+			Capabilities: &v1.Capabilities{
+				Drop: []v1.Capability{"ALL"},
+				Add:  []v1.Capability{"NET_BIND_SERVICE"},
+			},
+		}
+		pod.Spec.SecurityContext.RunAsGroup = &[]int64{uid}[0]
+		pod.Spec.SecurityContext.RunAsUser = &[]int64{uid}[0]
+		pod.Spec.SecurityContext.SupplementalGroups = []int64{controlPlanUsersAndGroups[kubeadmconstants.ServiceAccountKeyReadersGroupName]}
+	}
```

Comment:
@neolit123 - do you think we should also delete the users and group in the end if it is a dry-run? I didn't do that as I didn't think it was necessary.

Reply:
we should not create users / groups on the host if it's a dry run.
instead kubeadm should say "[dryrun] would create uid/gid foo"

there are some examples for that in the code base.